File: leetcode/leetcode_cpp/number-of-islands.cpp
#include <iostream>
#include <vector>
#include <cassert>
#include <stack>
#include <queue>
#include "../utils.h"

using namespace std;
/*
https://leetcode.com/problems/number-of-islands

Solution1. DSF inorder. Mark visited in place
time: o(n)
space: o(1)

Solution2. BSF. Mark visited in place
time: o(n)
space: o(1)

*/
class Solution {
public:
    void dfs(vector<vector<char>>& grid, int i, int j) {
        int n1 = grid.size();
        int n2 = grid[0].size();

        if (i < 0 || i >= n1 || j < 0 || j >= n2 || grid[i][j] == '0') return;
        
        grid[i][j] = '0';
        dfs(grid, i - 1, j);
        dfs(grid, i + 1, j);
        dfs(grid, i, j - 1);
        dfs(grid, i, j + 1);

    }

    int numIslands_1(vector<vector<char>>& grid) {
        int cnt = 0;
        for (int i=0; i<grid.size(); ++i) {
            for (int j=0; j<grid[0].size(); ++j) {
                if (grid[i][j] == '0') continue;
                cnt++;
                dfs(grid, i, j);
            }
        }
        return cnt;
    }

    int numIslands_2(vector<vector<char>>& grid) {
        if (!grid.size() || !grid[0].size()) return 0;
        int n1 = grid.size();
        int n2 = grid[0].size();

        int cnt = 0;
        for (int i=0; i<n1; ++i) {
            for (int j=0; j<n2; ++j) {
                if (grid[i][j] == '0') continue;

                cnt++;
                queue<pair<int,int>> q;
                q.push({i, j});
                
                while (!q.empty()) {
                    int qSize = q.size();
                    for (int k=0; k<qSize; ++k) {
                        int row = q.front().first;
                        int col = q.front().second;
                        q.pop();
                        
                        if (row < 0 || row >= n1 || col < 0 || col >= n2) continue;
                        if (grid[row][col] == '0') continue;

                        grid[row][col] = '0';
                        q.push({row-1, col});
                        q.push({row+1, col});
                        q.push({row, col-1});
                        q.push({row, col+1});
                    }
                }
            }
        }
        return cnt;
    }
    
    int numIslands(vector<vector<char>>& grid) {
        return numIslands_2(grid);
    }
};


int main()
{
    auto input1 = vector<vector<char>>();
    input1.push_back(vector<char>{'1', '1', '1', '1', '0'});
    input1.push_back(vector<char>{'1', '1', '0', '1', '0'});
    input1.push_back(vector<char>{'1', '1', '0', '0', '0'});
    input1.push_back(vector<char>{'0', '0', '0', '0', '0'});
    assert(Solution().numIslands(input1) == 1);
   
    auto input2 = vector<vector<char>>();
    input2.push_back(vector<char>{'1'});
    input2.push_back(vector<char>{'1'});
    assert(Solution().numIslands(input2) == 1);

    auto input3 = vector<vector<char>>();
    input3.push_back(vector<char>{'1', '1', '0', '0', '0'});
    input3.push_back(vector<char>{'1', '1', '0', '0', '0'});
    input3.push_back(vector<char>{'0', '0', '1', '0', '0'});
    input3.push_back(vector<char>{'0', '0', '0', '1', '1'});
    assert(Solution().numIslands(input3) == 3);

    return 0;
}

File: leetcode/leetcode_py/number-of-islands.py
'''
https://leetcode.com/problems/number-of-islands

Solution1. DSF inorder. Mark visited in place
time: o(n)
space: o(1)

Solution2. BSF. Mark visited in place
time: o(n)
space: o(1)
'''
class Solution:
    def numIslands_1(self, grid: List[List[str]]) -> int:
        n1 = len(grid)
        if n1 == 0: return 0
        n2 = len(grid[0])
        if n2 == 0: return 0

        def dfs(row, col):
            if row < 0 or col < 0 or row >= n1 or col >= n2 or grid[row][col] == '0':
                return
            grid[row][col] = '0'
            dfs(row-1,col)
            dfs(row+1,col)
            dfs(row,col-1)
            dfs(row,col+1)
            
        cnt = 0
        for i in range(n1):
            for j in range(n2):
                if grid[i][j] == '0':
                    continue
                cnt += 1
                dfs(i, j)
        return cnt

    def numIslands_2(self, grid: List[List[str]]) -> int:
        n1 = len(grid)
        if n1 == 0: return 0
        n2 = len(grid[0])
        if n2 == 0: return 0

        cnt = 0
        for i in range(n1):
            for j in range(n2):
                if grid[i][j] == '0':
                    continue
                cnt += 1
                q = [[i,j]]
                while q:
                    values = q.pop(0)
                    row = values[0]
                    col = values[1]
                    if row < 0 or col < 0 or row >= n1 or col >= n2 or grid[row][col] == '0':
                        continue
                    grid[row][col] = '0'
                    q.append([row-1,col])
                    q.append([row+1,col])
                    q.append([row,col-1])
                    q.append([row,col+1])
                
        return cnt

    def numIslands(self, grid: List[List[str]]) -> int:
        return self.numIslands_2(grid)
        